checkbox 컴포넌트를 구현한다.

al-bur opened the pull request:
checkbox 컴포넌트 구현했습니다. 어떻게 구현할지를 고민하고 삽질해서 조금 시간이 걸렸네요
콤피 덕분에 쉽게 구현할 수 있었습니다~

코드량이 많지 않아 리뷰할 게 많이 없을 것 같지만 그래도 가감없이 피드백 요청드립니다.

## 기능

- checkbox 구현
- checkbox storybook 구현

## 특이사항

- ...props로 모든 props들을 자유롭게 받을 수 있도록 해주었습니다.
- text와 함께 체크박스를 사용하는 경우도 있어서, children을 열어주었습니다.


close #13

Commit:
feat: 제어/비제어 방식으로 컨트롤 할 수 있도록 checkbox 컴포넌트 구현

Commit:
refactor: 체크박스 스토리북 수정

Commit:
refactor: styled-components로 수정 및 컴포넌트 수정

Commit:
refactor: 불필요한 코드 수정

Commit:
fix: storybook 오류 수정

frontend-jihyeok-um commented on packages/cart/src/components/Checkbox/Checkbox.stories.tsx:
```diff
@@ -0,0 +1,35 @@
+};
+
+export const DefaultChecked: Story = {
+  render: () => <Checkbox defaultChecked />
```
`defaultChecked` 와 `checked={false}`가 프로퍼티로 같이 오면 어떻게 될까 해서 해봤더니 false가 되는군요!

그 외에 checkbox 컴포넌트에서 사용하지 않아야할 프로퍼티들도 props로 받아지고 있는 것 같은데 props타입을 좀 수정하면 어떨까요?

al-bur replied:
`defaultChecked 와 checked={false}가 프로퍼티로 같이 오면 어떻게 될까 해서 해봤더니 false가 되는군요!`

이 부분은 수정하겠습니다!! 다른 ui 라이브러리를 참고해보니, defaultChecked를 우선시해서 처리를 해주는 것 같더라고요. 사실 이 부분은 어떻게 정하느냐에 따라 달라질 부분인 것 같긴 합니다.

al-bur replied:
예상하지 못했던 부분들 딱 잡아내주셔서 감사해요 해리~

frontend-jihyeok-um commented on packages/cart/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,22 @@
+type Props = PropsWithChildren<InputHTMLAttributes<HTMLInputElement>>;
+
+const Checkbox = forwardRef(function Checkbox(
+  { children, ...props }: Props,
```
```
 <Checkbox defaultChecked type="number" />
```

당연히 이런 식으로 사용하는 사람은 없겠지만(...)
Checkbox 컴포넌트에서 type 프로퍼티는 받을 필요 없어 보입니다!
실제로 저렇게 되면 컴포넌트가 동작하지 않는 버그가 있기도 하고요
props의 타입을 checkbox 컴포넌트에 맞게 조금 수정해주면 어떨까 싶네요!

al-bur replied:
그렇네요 그 부분 신경써서 재작성해보겠습니다! 예상하지 못했던 부분들 딱 잡아내주셔서 감사해요 해리~

al-bur replied:
type이라는 props는 type에서 받을 수 없도록 Omit 해줬어요

frontend-jihyeok-um commented on packages/cart/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,22 @@
+      <S.ImageWrapper>
+        <img src={checkImage} />
+      </S.ImageWrapper>
+      {children}
```
추가로 children으로 checkbox 옆의 문자를 받는 것도 저는 조금 의아하네요!
```tsx
 <Checkbox defaultChecked>체크박스</Checkbox>
```
보통 이렇게 쓰게되면 Checkbox안에 글씨가 온다고 이해하기 쉬운 것 같아요
`<button>버튼 텍스트</button>` 이렇게 작성하는 것처럼요! 
새로운 프로퍼티로 텍스트를 받는 게 좋지 않나 싶네요!

children을 사용하면 정말 타입이 넓어져서 이런 짓도 가능하고요!
```tsx
<Checkbox defaultChecked>
 <div style={{ width: '100px', height: '100px', backgroundColor: 'black' }}></div>
</Checkbox>
```
<img width="190" alt="스크린샷 2023-05-04 오후 7 35 23" src="https://user-images.githubusercontent.com/75592315/236181026-79997526-5baa-4cdc-ad66-21887128ba6f.png">

compy-ryu replied:
엇, 전 사실 해리 의견이랑 살짝 상충되는데요!

일반적으로 `Checkbox` 우측에 문자열을 표기하고 해당 체크박스가 무엇인지 표기하는 경우가 더 많다 느꼈습니더.
그리고 만약 `children`을 받지 않을 시 아래와 같이 매번 감싸주어야 되는데요.

```jsx
<label>
  <Checkbox />
  <Text>체크박스입니더</Text>
</label>
```

이렇게 되면 코드라인도 길어지고, Text의 스타일, checkbox랑 text 사이 간격 등
개발자, 디자이너에 따라 체크박스와 텍스트 간격, 폰트 크기/색상 등이 틀려져 디자인 일관성을 헤칠 수 있다 생각해서요.

디자인 시스템을 사용하는 목적이 일관성, 생산성 향상인데
이런 목적에는 children을 받는 것이 더 적합하지 않을까합니당.

예)
- 만약 체크박스를 해제했을 때 텍스트가 연하게 표기되어야 한다는 스펙이 추가된다면?
- 개발자마다 체크박스와 텍스트 사이 gap, margin을 다르게 설정한다면?

---

확장성도 고려한다면 Checkbox children을 열어주되
children 없이 사용하는 경우에 label 태그를 사용하지 않도록 같이 처리해주는게 어떨까하는 의견 드립니더!

al-bur replied:
두 분의 날카로운 분석 너무 감사드려요!! 저도 편리성 측면에서 콤피의 의견에 조금 더 동의를 하는 바입니다!!

> children 없이 사용하는 경우에 label 태그를 사용하지 않도록 같이 처리해주는게 어떨까하는 의견 드립니더!

이 부분은 좋은 것 같아요. children이 없다면, label도 필요 없으니깐요!


> 보통 이렇게 쓰게되면 Checkbox안에 글씨가 온다고 이해하기 쉬운 것 같아요

해리가 언급해주신 포인트도 생각해봐야할 것 같아요! 이 부분은 사용자들이 저희 라이브러리를 사용할 때 checkbox에 text를 넣고 싶으면 어떻게 해야하지?라는 고민이 생겼을 때 저희 docs를 보면 한 번에 방법을 알아챌 수 있도록 가독성있게 docs를 만들어서 해결해줄 수 있을 것 같아요

al-bur replied:
> children 없이 사용하는 경우에 label 태그를 사용하지 않도록 같이 처리해주는게 어떨까하는 의견 드립니더!
> > 이 부분은 좋은 것 같아요. children이 없다면, label도 필요 없으니깐요!

위 작업을 시도하다보니, 생각보다 작업이 길어질 것 같아서 일단은 children이 없어도 label 태그가 남아있는 쪽으로 진행했습니다

frontend-jihyeok-um reviewed (commented):
안녕하세요 앨버! 크게 changes가 없어서 많은 리뷰를 남길 건 없는 것 같네요!
과제하느라 바쁘실텐데 구현하느라 수고 많으셨습니다!
개인적인 의견을 리뷰로 남겨두었습니다~!

compy-ryu commented on packages/cart/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,22 @@
+
+type Props = PropsWithChildren<InputHTMLAttributes<HTMLInputElement>>;
+
+const Checkbox = forwardRef(function Checkbox(
```
```tsx
function Checkbox(props, ref: React.ForwardedRef<HTMLInputElement | null>) {

}

export default forwardRef(Checkbox)
```

별건 아니지만, 혹시 이건 어떨까유!?

al-bur replied:
변경했습니다!! 좋네요

44e980c

compy-ryu commented on packages/cart/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,22 @@
+  return (
+    <S.Label>
+      <S.Input type="checkbox" ref={ref} {...props} />
+      <S.ImageWrapper>
+        <img src={checkImage} />
+      </S.ImageWrapper>
```
ImageWrapper, img 태그로 하는 것보다는
div로 하나로 표현하고 background-image로 표현하는건 어떻게 생각하십니까!?

정말 별거 아닙니더만

element 하나로 표현할 수 있고
img 태그를 쓰시면 사용자가 오른쪽 마우스로 클릭하면 이미지 다운로드가 뜨고
혹은 체크박스를 드래그해서 움직이면 이미지가 끌려오고, 체크박스를 빠르게 더블 클릭하면 체크 이미지가 드래그 되어버립니다.

al-bur replied:
와우...너무 좋은 것 같아요! 반영했습니다ㅎㅎ 하나 배워갑니다~

fe0eca3

compy-ryu commented on packages/cart/src/components/Checkbox/styles.ts:
```diff
@@ -0,0 +1,40 @@
+import styled from 'styled-components';
+
+const Input = styled.input`
+  border: 0;
+  clip: rect(0px, 0px, 0px, 0px);
+  height: 1px;
+  width: 1px;
+  padding: 0;
+  margin: 0;
+  overflow: hidden;
+  white-space: nowrap;
+  position: absolute;
+`;
```
```suggestion
const Input = styled.input`
  display: none;
`;
```

`display: none`으로 해도 체크박스는 일 잘합니더!

al-bur replied:
반영 완료! e866ada

compy-ryu reviewed (approved):
형님! 늦어서 죄송합니다!

기능도 모두 동작하고 크게 리뷰할 부분은 없다보니께
작은 부분들도 리뷰해보았습니더..

바로 Approve 해두었고, 반영은 형님 편하신대로 부탁드립니더!
고생하셨습니다!!

Commit:
refactor: forwardref 타입 변경 및 형식 변경

Commit:
refactor: 체크박스 이미지 관련 css 수정

- 콤피 피드백 반영: https://github.com/dry-frontend/dry-design-system/pull/17\#discussion_r1185921348

Commit:
refactor: 체크박스내 input css 불필요없는 특성들 교체

- 피드백 반영: https://github.com/dry-frontend/dry-design-system/pull/17\#discussion_r1185922222

Commit:
feat: label에 대한 props를 받을 수 있도록 기능 추가

Commit:
refactor: 타입으로 type attribute 사용하지 못하도록 변경

Commit:
fix: merge conflict 해결
